### ISSUE TITLE
Please add a switch to enable or disable auto live-preview update + button to manually update in prev. mode

### ISSUE THREAD

**Issue by bencresty** [craftcms/cms#7941]
### Description

I really love the live update and it's great for customers to edit content with a live preview. Currently I am working on projects that have a lot of animations though and I see customers/myself editing everything with live preview open most of the time. That's great, because it's very visual and I like it, but it's not so great when all animations are triggered in the preview every time because it reloads/refreshes on basically every character we type. I understand this could be nice in some situations, like building static websites, but it's not so great when working on projects that have a lot of animations and don't load instantly as we see stuff reloading, flashing and animating all the time while just typing.

Yes, we could add a special preview mode in our application being shown in the preview, but than again it still needs time to load stuff with ajax and all and there's no need to refresh on each character we type. It's too much for our needs and disturbes the workflow now as it's constantly keeping flashing/reloading/animating while typing, making us not only feel rushed and disturbed, we could also not see the preview during these moments as it's constantly busy refreshing.

Also it's not great if we are supposed to build a special preview mode in each headless app just to prevent this situation. As that's not only more work (and so takes time and money), but is also more code in the front end that's only there for this.

It would be nice if we had a switch on the edit screen next to the preview page to switch automatic live updating temporarly off. And to have some button next to it to manually update the preview when we need it. That would not rush the content editor to type everything as quickly as possible, just to not let the preview refresh all the time (I find myself doing that a lot and to be honest, it gives a very rushed feeling to type quickly just to prevent the preview from updating/flashing/animating constantly).

Really hope you could consider this, it would help staying relaxed and not rushed while editing with the preview window open! Thanks!

**Comment by brandonkelly** [comment]
We’ll think about this, but in the meantime, you could opt out of showing the animations by adding this to your `<body>` tag:

```twig
<body {% if craft.app.request.isPreview %}class="no-animations"{% endif %}>
```

Then from your CSS, add `body:not(.no-animations)` to the beginning of each of your animation selectors. (Or if it’s in JS, check for the class there.)

**Comment by bencresty** [comment]
> We’ll think about this, but in the meantime, you could opt out of showing the animations by adding this to your `<body>` tag:
> 
> ```
> <body {% if craft.app.request.isPreview %}class="no-animations"{% endif %}>
> ```
> 

I understand. I do it with headless cp and url param in the url to switch to a preview mode. But it's a lot of extra work on each project and redundant extra code in front end, which we optimize for fastest loading. Also turning off animations isn't the only thing; it still needs to load data via Ajax etc. so loading the preview just isn't instant, so when typing a char in the CP we see the loading spinner again in the preview and keep on seeing it reloading on each character until the last character we've typed. It just makes you feel restless and hurried when editing AND you don't see a preview in the meantime of the previous state (which in a lot of times is enough to edit a text on the fly).

Most times we only want to see a result AFTER changing. While typing or editing images or whatever we normally aren't looking at the preview anyhow, because we need our attention on the edit screen.

We could say an update on leaving a field would be the right moment, but that doesn't work with, for example, rich content fields, because we stay in the field. Also a longer delay wouldn't help either as than it would be too long to wait.

That's why I think a great solution would be (and for everybody to pick his/her own workflow) to be able to (temporarly) switch off the auto update timer and have a button next to it to update the preview when wanted.
That also has a great other advantage over auto update: we could use the app inside the preview without it getting 'reset' on each edit. That's for instance very welcome on Google Maps where you want to zoom in to find a location for instance and want to type that location in a field of the CP. Than it's pretty frustrating when the map resets its zoom level on each character you type, because you'd like to see the values while you type. This is just a real life example (in this case, when only needing this it could be solved differently, but this is just an example, I've had several cases like this, bottom line is that sometimes we don't want the preview to be refreshed because it's resetting the app state).

> Then from your CSS, add `body:not(.no-animations)` to the beginning of each of your animation selectors. (Or if it’s in JS, check for the class there.)

I understand CSS. The animations we make are a little more complicated than this though and reach far outside CSS and use several techniques. We're not talking about a simple button animation here and not even only animations in interface, but full advanced animations everywhere, in both 2D/3D with math etc. Sure, we can all turn things off, but it's not always that easy and influences everything and need a lot of extra testing.
So to come back to my first point;
- That's way more work and takes a lot of extra time to program and test while it's only to prevent auto-update of a preview in the cms
- It's still not solving the fact that the app-state of the app inside the preview changes on every character we type, which makes the content-editor need to go back to the state he wants to be in on each edit, while this could be simply avoided by turning off the auto reload of the app
- It's still not solving the fact that even if animations are off in preview mode (and we had to put a lot of extra effort and extra code to accomplish that) the app still needs to reload data by ajax, graphics, sounds etc at start and so we always see a loading spinner instead of the app for a little while while just typing some characters
- It's still not solving that we wished to be able to see the preview while typing, even if it's not the actual update. This would be great, because the content editor can than change texts, images, etc. with this preview in mind. And only need to see the actual result after a click he does when he things everything has been edited as wanted.

Bottom line is; it's not a regular static site with some css animations, but a full blown app with states that animate everything, not only the interface.

Hope this makes sense!

**Comment by thupsi** [comment]
@bencresty If you are working on a draft, the draft is autosaved with each edit, so if you have a "view" window/tab open (have clicked "view" while on a draft edit page) you can refresh the seperate window/tab on demand. Only one click needed. Wouldn't this suffice for what you are trying to accomplish or am I missing something?

**Comment by brandonkelly** [comment]
Just added a “Refresh” button, for preview targets whose “Auto-refresh” (formerly “Refresh”) setting is disabled.

**Comment by bencresty** [comment]
@brandonkelly thanks! I am not familiar with the formerly 'Refresh' setting and so far couldn't find info on it;
Does this mean it will be possible to switch the auto-refresh temporarly on and off on the edit page / preview? Or is there one general setting to turn on/off auto-refresh everywhere in general.php, or perhaps switchable per entry type?

**Comment by brandonkelly** [comment]
@bencresty If you are running Craft Pro, there is a “Refresh” checkbox column in your sections’ Preview Targets tables (in the section settings). If that is unchecked, the preview iframe will never be automatically refreshed as content is updated.

That was originally added with HMR and tools like Gatsby & Gridsome in mind, where the page is capable of updating itself without needing a full reload. But even in those cases, it wouldn’t hurt to give authors the ability to click a Refresh button to force a full reload.
